### PR TITLE
plugins/cmake-tools: init

### PR DESCRIPTION
**Pull request by nfelber**
This PR adds the cmake-tools plugin to nixvim.

A substantial part of this plugin's default config sets options for external plugins such as dap or overseer (optional), therefore I didn't create nixvim options for these and used generic attribute sets instead as they seem to be beyond the scope of cmake-tools itself and aren't exhaustive. Please let me know if this is not the right approach - one issue I noticed is that large attribute set default values don't render well in the docs.

### CODE REVIEW

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [inline plugins/languages/cmake-tools.nix:288]: comments in `settingsExample` won't show up in the docs. Most of these should probably be converted to enum types of moved to description text as appropriate, unless they're meant as internal comments for nixvim maintainers.

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [inline plugins/languages/cmake-tools.nix:244]: @traxys @GaetanLepage if `[ "⠋" "⠙" "⠹" "⠸" "⠼" "⠴" "⠦" "⠧" "⠇" "⠏" ]` is the [plugin's default](https://github.com/Civitasv/cmake-tools.nvim?tab=readme-ov-file#balloon-configuration), should _our_ default value just be `null`? How do we normally handle defaults like this?

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [inline plugins/languages/cmake-tools.nix:111]: Would some of these `mkAttributeSet` options be better off as submodues? How do we normally handle this while sticking to [RFC 42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md)?

File: plugins/languages/cmake-tools.nix
Comment (nfelber) [reply plugins/languages/cmake-tools.nix:288]: Thanks for the review; these comments are copied from the default cmake-tools configuration upstream. I wasn´t aware that they wouldn't show up in the docs - then there is no point leaving them for nixvim maintainers. I'll just wait for further input to your other comments before proceeding, and accordingly remove them or move them to option descriptions.

Comment (MattSturgeon) [reply plugins/languages/cmake-tools.nix:288]: Personally, I think we should use enum options when upstream document a list of acceptable values.

Unless that list is likely to change frequently; that'd become a maintainance burden.

Comment (GaetanLepage) [reply plugins/languages/cmake-tools.nix:288]: > Personally, I think we should use enum options when upstream document a list of acceptable values.
> 
> Unless that list is likely to change frequently; that'd become a maintainance burden.

100% agree.

**Comment by GaetanLepage** [comment]
I haven't read everything, but keep in mind that it is absolutely not mandatory to package every single option.
This one has a lot of options, so we might be better of selecting the most valuable ones.

As you have already went through the work, we can also keep them all, but keep in mind that it is not a requirement.

**Comment by nfelber** [comment]
I'm not sure I understand what would be best to do. On one hand I totally agree that not every option should be packaged in nixvim. This is partly why I left several options as attribute sets instead of composite options. On the other hand, I also agree that options listed upstream would be better packaged as enum values whenever possible, but I don't know how to properly satisfy both constraints.

For example, this is the default upstream `cmake_dap_configuration` value:

```lua
cmake_dap_configuration = { -- debug settings for cmake
  name = "cpp",
  type = "codelldb",
  request = "launch",
  stopOnEntry = false,
  runInTerminal = true,
  console = "integratedTerminal",
}
```

I packaged it as a simple attribute set for the following reasons:

- To limit the number of available options as discussed above.
- This config is directly passed to nvim-dap (if used), and is not immediately related to cmake-tools.
- While some attributes like `name`, `type` and `request` are required by nvim-dap, the others are either optional or specific to the debug adapter ("codelldb" in this case), and listing them all would be very difficult to maintain.
- If I were to add proper options for the `name`, `type` and `request` attributes (strings and enum), I don't think arbitrary additional attributes would be supported anymore  (?).

A similar reasoning applies to the `cmake_executor.default_opts` and `cmake_runner.default_opts` options.

Hopefully this clarifies the confusion, please let me know what you think and how to go about it.

File: plugins/languages/cmake-tools.nix
Comment (GaetanLepage) [reply plugins/languages/cmake-tools.nix:244]: This is the right approach. The default provided to the `helpers.defaultNullOpts.*` helpers is the one of the plugin.
The actual default value of the nixvim option will always be `null`.

**Comment by MattSturgeon** [comment]
@nfelber do you need any help with this one and are you still working on it?

If you prefer, one of us could continue what you started when we get chance. Otherwise we can answer any specific questions (here or on matrix) and do another round of reviews.

**Comment by nfelber** [comment]
@MattSturgeon yes, I'd like to continue working on it. I haven't made progress in a while because I'm unsure what direction to take to improve the packaging, as described in my comment above.

If you could have a look, that would be very helpful. And feel free to let me know if anything is unclear.

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [inline plugins/languages/cmake-tools.nix:168]: Again, prefer not using composite options, but you can if you think the description is necessary:

```nix
cmake_runner = {
```

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [inline plugins/languages/cmake-tools.nix:162]: Same as `default_opts`, make this into a set of options if you think it's worth keeping.

If it's the same options that `default_opts` uses you can declare them in a `let in` block and reuse them, or perhaps make a `submodule` type that can be reused.

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [inline plugins/languages/cmake-tools.nix:95]:
```suggestion
    cmake_dap_configuration =
      helpers.defaultNullOpts.mkAttrsOf lib.types.anything {
          name = "cpp";
          type = "codelldb";
          request = "launch";
          stopOnEntry = false;
          runInTerminal = true;
          console = "integratedTerminal";
        } ''
        Dap configuration for cmake.
      '';
```

- We no longer quote plugin-defaults
- `attrsOf anything` is preferred over `attrs`
- If `name`, `type`, `request`, etc are well known options, you could make them actual options instead. Or define a submodule type to use here.

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [inline plugins/languages/cmake-tools.nix:68]:
```suggestion
        short =
          helpers.defaultNullOpts.mkAttrsOf lib.types.anything { show = true; }
```

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [inline plugins/languages/cmake-tools.nix:64]: Again:

```suggestion
    cmake_variants_message = {
```

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [inline plugins/languages/cmake-tools.nix:35]:
```suggestion
      helpers.defaultNullOpts.mkAttrsOf lib.types.anything {
        "-DCMAKE_EXPORT_COMPILE_COMMANDS" = 1;
      }
```

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [inline plugins/languages/cmake-tools.nix:40]:
```suggestion
    cmake_build_options = helpers.defaultNullOpts.mkAttrsOf lib.types.anything { } ''
```

File: lib/maintainers.nix
Comment (MattSturgeon) [inline lib/maintainers.nix:1]: The maintainers commit should be titled:

> `maintainers: add NathanFelber`

**Review by MattSturgeon — CHANGES_REQUESTED** [review]
I hope this is helpful. I've gone over everything quickly, so probably missed some stuff that'll be picked up in later reviews.

Note, you'll also need to run `nix fmt` in order for CI to pass.

Let me know here or on matrix if you have any questions of if you'd like me to push up any changes to your branch.

File: tests/test-sources/plugins/languages/cmake-tools.nix
Comment (nfelber) [reply tests/test-sources/plugins/languages/cmake-tools.nix:6]: Yes, this is correct. Thanks for pointing it out.

File: tests/test-sources/plugins/languages/cmake-tools.nix
Comment (nfelber) [reply tests/test-sources/plugins/languages/cmake-tools.nix:111]: Yes, it is indeed an attrs type, I had no idea how to implement it - this is exactly what I need. Isn't it `strategy.__unkeyed-1` however (from looking at `to-lua.nix`)?

Comment (MattSturgeon) [reply tests/test-sources/plugins/languages/cmake-tools.nix:111]: Yes, that was a typo on my end

File: plugins/languages/cmake-tools.nix
Comment (nfelber) [reply plugins/languages/cmake-tools.nix:235]: That is cleaner indeed. I probably thought every attribute had to be an option for some reason.

**Comment by nfelber** [comment]
Thank you very much for your detailed feedback. It is incredibly helpful. I'll continue going through the remaining suggestions when I get time later this week.

File: plugins/languages/cmake-tools.nix
Comment (nfelber) [reply plugins/languages/cmake-tools.nix:95]: In this case, `name`, `type` and `request` are well known required options while the others are optional and depend on the debug adapter being used (e.g. `codelldb`). It would be nice to package the first three but also let the user add arbitrary attributes, and I'm not sure how to properly do that. mkCompositeOption doesn't seem to take a plugin default, and mkAttrsOf anything doesn't seem to take child options. Any idea?

File: plugins/languages/cmake-tools.nix
Comment (nfelber) [reply plugins/languages/cmake-tools.nix:229]: Yes, thinking twice about it, `default_opts` are probably not worth keeping. I'll have a closer look, but I don't think they are supposed to be modified anyways as they are part of the `cmake-tools.nvim` default config and are overridden by the user-defined `opts`. And they are also specific to optional external plugins, therefore out of the scope of `cmake-tools.nvim` as far as option packaging is concerned.

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [reply plugins/languages/cmake-tools.nix:95]: > It would be nice to package the first three but also let the user add arbitrary attributes, and I'm not sure how to properly do that

The `settings` option has a "freeform" type (`attrsOf anything`), so this is the default behaviour, unless we add additional restrictions via option types.

If that's what you want here then all you need to do is:

```nix
cmake_dap_configuration = {
  name = /*mkOption*/;
  type = /*mkOption*/;
  request = /*mkOption*/;
};
```

If this needs to be done in a re-usable way, then there's a few ways to do this. The simplest would be to define the option attrs in a let block and re-use it that way.

With this approach you could also define "extra" options using the `commonOpts // {/* extraOpts */}` pattern (if you wanted to).

---

It's also possible to do this kinda thing with submodules, however `mkCompositeOption` is probably not the best way to define a submodule option, because it doesn't set a freeform type and AFAIK submodules don't inherit the parent's freeform type.

A submodule option _would_ allow you to define things like `example`, `description` and `pluginDefault` though.

I'd suggest ignoring submodules for now, focus on getting something simple working and then revisit the idea later if you're not satisfied with the end result.

File: tests/test-sources/plugins/languages/cmake-tools.nix
Comment (nfelber) [reply tests/test-sources/plugins/languages/cmake-tools.nix:117]: As I've decided to not package this option, I'll stick with `__raw` in the test file.

File: plugins/languages/cmake-tools.nix
Comment (nfelber) [reply plugins/languages/cmake-tools.nix:95]: I see, thank you for the clarification. Then I'll go the simplest route and only declare these three options as you suggest.

**Comment by MattSturgeon** [comment]
@nfelber looks like you've rebased `main` onto your branch instead of your branch onto `main`

**Comment by nfelber** [comment]
Oh, so that's what happened. My bad, I'll try to fix it later today.

**Comment by nfelber** [comment]
@MattSturgeon Looks like it's now fixed. I think I integrated all your suggestions into the PR and it should be ready for review. However, some of the checks are failing and I can't test locally as my system (16 GB RAM + 16 GB swap) runs out of memory while running `nix flake check --all-systems`  and kills the process. Is that normal?

**Comment by MattSturgeon** [comment]
> @MattSturgeon Looks like it's now fixed. I think I integrated all your suggestions into the PR and it should be ready for review. However, some of the checks are failing and I can't test locally as my system (16 GB RAM + 16 GB swap) runs out of memory while running `nix flake check --all-systems` and kills the process. Is that normal?

Try entering `nix develop` (to get a devshell) and using the `tests` command, instead of running `nix flake check`.

You can also try omitting `--all-systems` as that'll make the tests _much_ heavier.

For additional context, we're having issues with our CI currently, so ignore any ci failure complaining about "broke pipe".

Testing locally shouldn't have that issue though.

Also: @traxys recently managed to dramatically improve the memory usage of the tests, you should benefit from that now you've rebase.

Before the tests needed ~32gb ram, now they use around 3gb or less.

**Comment by MattSturgeon** [comment]
The fact the CI is passing on most systems, and failing with "broken pipe" on two tells me the tests are passing for this PR.

I'll review again when I have chance, hopefully over the weekend!

Thanks for your efforts!

File: plugins/languages/cmake-tools.nix
Comment (GaetanLepage) [inline plugins/languages/cmake-tools.nix:15]: This should be dealt with at the nixpkgs level.

File: plugins/languages/cmake-tools.nix
Comment (GaetanLepage) [inline plugins/languages/cmake-tools.nix:16]: Could you please add a `settingsExample` section as well ?

**Comment by nfelber** [comment]
> Try entering `nix develop` (to get a devshell) and using the `tests` command, instead of running `nix flake check`.
> 
> You can also try omitting `--all-systems` as that'll make the tests _much_ heavier.

I see, thanks for the tips and context. I'll give it another try tomorrow then!

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [reply plugins/languages/cmake-tools.nix:16]: Note for context: I'd previously suggested removing the example as it just contained the plugin defaults.

If we can come up with a useful non-default example that'd be ideal though.

It could be extensive, or it could just set one or two "important" or frequently configured options.

Maybe try searching `lang:lua "cmake-tools"` accross GitHub to see real world examples, or check if upstream has any examples in its docs?

**Comment by nfelber** [comment]
> The fact the CI is passing on most systems, and failing with "broken pipe" on two tells me the tests are passing for this PR.
> 
> I'll review again when I have chance, hopefully over the weekend!

Awesome, thank you then!

File: plugins/languages/cmake-tools.nix
Comment (nfelber) [reply plugins/languages/cmake-tools.nix:15]: The dependency on `plenary.nvim` you mean? I can have a look when I get some time.

File: plugins/languages/cmake-tools.nix
Comment (nfelber) [reply plugins/languages/cmake-tools.nix:16]: Sure, I'll add one with non-default values.

File: plugins/languages/cmake-tools.nix
Comment (GaetanLepage) [reply plugins/languages/cmake-tools.nix:15]: https://github.com/NixOS/nixpkgs/pull/328534

File: plugins/languages/cmake-tools.nix
Comment (nfelber) [reply plugins/languages/cmake-tools.nix:16]: I added a basic example. It only slightly deviates from the default configuration, but in what I consider could be a useful (or at least usable) way. I wasn't able to find examples online or upstream unfortunately, but the `cmake_dap_configuration` part comes from [the nvim-dap wiki](https://github.com/mfussenegger/nvim-dap/wiki/C-C---Rust-(via--codelldb)).

**Review by GaetanLepage — CHANGES_REQUESTED** [review]
Good !
Could you please squash your commits so that you have only the two following:
- `helpers/maintainers: add NathanFelber`
- `plugins/cmake-tools: init`

**Comment by nfelber** [comment]
Done. Let me know if there's anything else I can do!

**Review by GaetanLepage — APPROVED** [review]
LGTM !
Thanks @nfelber and the reviewers for this nice work !

**Review by GaetanLepage — CHANGES_REQUESTED** [review]
We need to wait for the next flake update to get the `plenary` dependency in.
In the meantime, you can remove the relevant line from your patch.

File: plugins/languages/cmake-tools.nix
Comment (MattSturgeon) [reply plugins/languages/cmake-tools.nix:15]: We should have https://github.com/NixOS/nixpkgs/pull/328534 in nixos-unstable within a day or two and our CI should also be back online soon.

Once we've updated our lockfile you'll be able to rebase and remove this line. I think we can wait for that before merging.

**Review by MattSturgeon — APPROVED** [review]
No issues from me, but lets hold off until we can run our CI checks and drop the plenary line.

Thanks for your hard work and persistence on this!

**Comment by nfelber** [comment]
Yes, no problem. I just amended my last commit with the relevant line removed.
And thank you for taking the time to guide me through the process!

**Review by GaetanLepage — APPROVED** [review]
All good for me !
Thanks @nfelber

**Comment by MattSturgeon** [comment]
@mergifyio queue

**Comment by MattSturgeon** [comment]
@mergifyio refresh

**Comment by MattSturgeon** [comment]
@mergifyio dequeue

**Pull request merged**
